graphql_compiler/compiler/emit_sql.py

```diff
@@ -876,25 +902,41 @@ def unfold(self):
         self._aliases[subquery_alias_key] = fold_subquery
         self._from_clause = from_cls
 
-        # Ensure references to the outer vertex table after the Unfold refer to a totally new
-        # copy of the outer vertex table. Otherwise references would select columns from the
-        # copy of that table found inside the fold subquery.
-        self._aliases[(self._current_location.at_vertex().query_path, None)] = outer_vertex
-
         # clear the fold from the compilation state
         self._current_fold = None
         self._fold_vertex_location = None
         self._current_alias = outer_vertex
+        self._current_location = self._current_location.base_location
 
-    def mark_location(self):
+    def mark_location(self, next_ir_block):
         """Execute a MarkLocation Block."""
-        # If the current location is the beginning of a fold, the current alias
-        # will eventually be replaced by the resulting fold subquery during Unfold.
-        self._aliases[
-            (self._fold_vertex_location.base_location.query_path,
-             self._fold_vertex_location.fold_path)
-            if self._current_fold is not None else (self._current_location.query_path, None)
-        ] = self._current_alias
+        # If we are in a fold and the next block either backtracks or unfolds, then we have
+        # traversed to the deepest vertex of the fold and need to call visit_output_vertex on
+        # the current vertex.
+        backtracks_next = isinstance(next_ir_block, blocks.Backtrack)
+        unfolds_next = isinstance(next_ir_block, blocks.Unfold)
+        if self.is_in_fold() and (unfolds_next or backtracks_next):
+            is_deepest_fold_vertex = True
+        else:
+            is_deepest_fold_vertex = False
+
+        # Visit output vertex if the current location is the deepest vertex in the fold
+        if is_deepest_fold_vertex:
+            # If we are at the deepest vertex in the fold, then the current location is the output
+            # vertex of the fold and we must visit it now to select its output columns.
+            self._current_fold.visit_output_vertex(self._current_alias,
+                                                   self._current_location,
+                                                   self._all_folded_fields)
```

Comment:
Since `visit_output_vertex` should be called exactly once, can we assert that? Better yet, why not have a `visit_vertex` instead of `visit_output_vertex` that's expected to be called on every vertex inside the fold. That will allow us to move all of this fold logic (checking for outputs) inside the `SQLFoldObject`, which is the class that we made to contain all the fold logic. Inside that method we can assert we've only found one vertex that actually has outputs. That will also remove the need to pass the unintuitive `next_ir_block` in here.

Reply:
I believe we would still need to determine whether this is an output vertex while in the context of the CompilationState because the SQLFoldObject has no way of knowing without any additional information whether the vertex is an output vertex.

At best, we could change the logic to have a visit_vertex method that receives a flag indicating this an output vertex when called from inside this if statement. 

No matter what, I don't think we get away with having some sort of version of this if-logic occurring in the CompilationState.

Reply:
Also, note that we do assert that `visit_output_vertex` is called exactly once inside of that function. But technically that check only works if we assume that we never pass in `None` as the `output_alias`. I can change the error checking to catch that case. That way we will always be ensured only one call to `visit_output_vertex` occurs.